sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -1047,6 +1047,53 @@ case class ToDate(child: Expression) extends UnaryExpression with ImplicitCastIn
 }
 
 /**
+ * Parses a column with a format to a timestamp.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(timestamp_str, fmt) - Parses the `timestamp` expression " +
+    "with the `fmt` expression.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_('2016-12-31', 'yyyy-MM-dd');
+       2016-12-31
+             """)
+case class ParseToDate(left: Expression, right: Expression, child: Expression)
+  extends RuntimeReplaceable with ImplicitCastInputTypes {
+
+  def this(left: Expression, format: Expression) = {
+    this(left, format, ToDate(new ParseToTimestamp(left, format)))
+  }
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType)
+  override def dataType: DataType = DateType
+  override def flatArguments: Iterator[Any] = Iterator(left, right)
+  override def sql: String = s"$prettyName(${left.sql}, ${right.sql})"
+}
+
+/**
+ * Parses a column with a format to a timestamp.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(timestamp, fmt) - Parses the `timestamp` expression with the `fmt` expression.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_('2016-12-31', 'yyyy-MM-dd');
+       2016-12-31 00:00:00.0
+  """)
+case class ParseToTimestamp(left: Expression, right: Expression, child: Expression)
+  extends RuntimeReplaceable with ImplicitCastInputTypes {
+
+  def this(left: Expression, format: Expression) = {
+  this(left, format, Cast(UnixTimestamp(left, format), TimestampType))
```

Comment:
Now if I remove the cast call and turn this function into:

```scala
  def this(left: Expression, format: Expression) = {
  this(left, format, UnixTimestamp(left, format))
}
```

I will end up with
```scala
scala> import org.apache.spark.sql.functions._
import org.apache.spark.sql.functions._

scala>

scala> //////////// TIME INFO

scala> spark.range(1).select(to_timestamp(lit("2016-12-31"), "yyyy-MM-dd")).collect()
res0: Array[org.apache.spark.sql.Row] = Array([1970-01-01 11:05:19.476746])

scala> // expected is what the output of the below command

scala> spark.range(1).select(to_timestamp(lit("2016-12-31"), "yyyy-MM-dd").cast("timestamp")).collect()
res1: Array[org.apache.spark.sql.Row] = Array([null])

scala>

scala>

scala> //////////// DATE INFO

scala> spark.range(1).select(to_date(lit("2016-12-31"), "yyyy-MM-dd")).collect()
res2: Array[org.apache.spark.sql.Row] = Array([null])

scala> // expected is the output of the below command

scala> spark.range(1).select(to_date(to_timestamp(lit("2016-12-31"), "yyyy-MM-dd"))).collect()
res3: Array[org.apache.spark.sql.Row] = Array([null])
```